Subject: [PATCH] gitweb: optimize Encoding use

Make sure the fallback encoding is only looked up exactly once
rather than every time we call to_utf8 and the input string is
not UTF-8 or US-ASCII.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>
---
 gitweb/gitweb.perl | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index ada60eb8..2e431385 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -805,6 +805,20 @@ sub evaluate_gitweb_config {
 	read_config_file($GITWEB_CONFIG_SYSTEM);
 }
 
+our $encode_object;
+
+sub evaluate_encoding {
+	my $requested = $fallback_encoding || 'ISO-8859-1';
+	my $obj = Encode::find_encoding($requested) or
+		die_error(400, "Requested fallback encoding not found");
+	if ($obj->name eq 'iso-8859-1') {
+		# Use Windows-1252 instead as required by the HTML 5 standard
+		my $altobj = Encode::find_encoding('Windows-1252');
+		$obj = $altobj if $altobj;
+	}
+	$encode_object = $obj;
+}
+
 # Get loadavg of system, to compare against $maxload.
 # Currently it requires '/proc/loadavg' present to get loadavg;
 # if it is not present it returns 0, which means no load checking.
@@ -1416,6 +1430,7 @@ sub _reset_globals {
 
 sub run {
 	evaluate_gitweb_config();
+	evaluate_encoding();
 	evaluate_git_version();
 	my ($ml, $mi, $bu, $hl, $subroutine) = ($my_url, $my_uri, $base_url, $home_link, '');
 	$subroutine .= '$my_url = $ml;' if defined $my_url && $my_url ne '';
@@ -1681,7 +1696,7 @@ sub to_utf8 {
 	if (utf8::is_utf8($str) || utf8::decode($str)) {
 		return $str;
 	} else {
-		return decode($fallback_encoding, $str, Encode::FB_DEFAULT);
+		return $encode_object->decode($str, Encode::FB_DEFAULT);
 	}
 }
 
---
